apps/src/lib/config/genesis/transactions.rs

```diff
@@ -632,108 +570,85 @@ impl SignedTransferTx {
-        })
+impl<T> Signed<T> {
+    /// Instantiate data to be signed.
+    pub const fn new(data: T) -> Self {
+        Self {
+            data,
+            signatures: BTreeMap::new(),
+        }
     }
 
-    /// The signable data. This does not include the phantom data.
-    fn data_to_sign(&self) -> Vec<u8> {
-        [
-            self.token.serialize_to_vec(),
-            self.source.serialize_to_vec(),
-            self.target.serialize_to_vec(),
-            self.amount.serialize_to_vec(),
-        ]
-        .concat()
+    /// Return the inner wrapped `T`.
+    pub fn into_inner(self) -> T {
+        let Signed { data, .. } = self;
+        data
     }
 
-    /// Sign the transfer.
-    ///
-    /// Since we denominate amounts as part of validation, we can
-    /// only verify signatures on [`SignedTransferTx`]
-    /// types. Thus we only allow signing of [`TransferTx<Unvalidated>`]
-    /// types.
-    pub fn sign(self, key: &common::SecretKey) -> SignedTransferTx {
-        let sig = standalone_signature::<_, SerializeWithBorsh>(
-            key,
-            &self.data_to_sign(),
-        );
-        SignedTransferTx {
-            data: self,
-            signature: StringEncoded { raw: sig },
+    /// Sign the underlying data and add to the list of signatures.
+    pub async fn sign(&mut self, keys: &[common::SecretKey])
+    where
+        T: BorshSerialize + TxToSign,
+    {
+        for sk in keys {
+            self.signatures.insert(
+                StringEncoded::new(sk.ref_to()),
+                StringEncoded::new(
+                    standalone_signature::<_, SerializeWithBorsh>(
+                        sk,
+                        &self.data.tx_to_sign(),
+                    ),
+                ),
+            );
         }
     }
-}
-
-pub type SignedBondTx = Signed<BondTx<Unvalidated>>;
 
-impl SignedBondTx {
-    /// Verify the signature of `BondTx`. This should not depend
-    /// on whether the contained amount is denominated or not.
-    ///
-    /// Since we denominate amounts as part of validation, we can
-    /// only verify signatures on [`SignedBondTx`]
-    /// types.
+    /// Verify the signatures of the inner data.
     pub fn verify_sig(
         &self,
-        pk: &common::PublicKey,
-    ) -> Result<(), VerifySigError> {
-        let Self { data, signature } = self;
-        verify_standalone_sig::<_, SerializeWithBorsh>(
-            &data.data_to_sign(),
-            pk,
-            signature,
-        )
+        pks: &[common::PublicKey],
+        threshold: u8,
+    ) -> Result<(), VerifySigError>
+    where
+        T: BorshSerialize + TxToSign,
+    {
+        let Self { data, signatures } = self;
+        if pks.len() > u8::MAX as usize {
+            eprintln!("You're multisig is too facking big");
+            return Err(VerifySigError::TooGoddamnBig);
+        }
+        let mut valid_sigs = 0;
+        let tx_to_sign = data.tx_to_sign();
+        for pk in pks {
```

Comment:
Is there a reason we need to reimplement all of this logic again? Can we just check the validity predicate that multisig accounts already have?